rust_src/src/obarray.rs

```diff
@@ -58,6 +58,37 @@ impl LispObarrayRef {
         }
     }
 
+    /// Ensure that we have a valid obarray.
+    pub fn check(self) -> Self {
+        // We don't want to signal a wrong-type error when we are shutting
+        // down due to a fatal error and we don't want to hit assertions
+        // if the fatal error was during GC.
+        if unsafe { fatal_error_in_progress } {
+            return self;
+        }
+
+        // A valid obarray is a non-empty vector.
+        let v = self.0.as_vector();
+        if v.map_or(0, LispVectorRef::len) == 0 {
+            // If Vobarray is now invalid, force it to be valid.
+            if unsafe { globals.Vobarray }.eq(self.0) {
+                unsafe { globals.Vobarray = initial_obarray };
+            }
+            wrong_type!(Qvectorp, self.0);
+        }
+
+        self
+    }
+
+    pub fn get(&self, idx: usize) -> LispSymbolRef {
+        LispObject::from(self).as_vector_or_error().get(idx).into()
+    }
+
+    pub fn set(&mut self, idx: usize, item: LispObject) {
```

Comment:
Perhaps this could be a generic function on `item`. The first line of the method then calls `into` on `item`.
```suggestion
    pub fn set<O: Into<LispObject>>(&mut self, idx: usize, item: O) {
```

Reply:
Ooh, great idea, all this casting can be verbose. Thanks!